lib/BuildSystem/BuildSystemFrontend.cpp

```diff
@@ -285,79 +287,214 @@ class BuildSystemFrontendExecutionQueueDelegate
 };
 
 struct BuildSystemFrontendDelegateImpl {
-
-  /// The status of delegate.
-  enum class Status {
-    Uninitialized = 0,
-    Initialized,
-    InitializationFailure,
-    Cancelled,
-  };
-
   llvm::SourceMgr& sourceMgr;
-  const BuildSystemInvocation& invocation;
-  
+
   StringRef bufferBeingParsed;
   std::atomic<unsigned> numErrors{0};
   std::atomic<unsigned> numFailedCommands{0};
 
   BuildSystemFrontendExecutionQueueDelegate executionQueueDelegate;
 
-  BuildSystemFrontend* frontend = nullptr;
-  BuildSystem* system = nullptr;
+  BuildSystemFrontendImpl* frontend = nullptr;
 
   /// The set of active command output buffers, by process handle.
   llvm::DenseMap<uintptr_t, std::vector<uint8_t>> processOutputBuffers;
 
   /// The lock protecting `processOutputBuffers`.
   std::mutex processOutputBuffersMutex;
-  
-  BuildSystemFrontendDelegateImpl(llvm::SourceMgr& sourceMgr,
-                                  const BuildSystemInvocation& invocation)
-      : sourceMgr(sourceMgr), invocation(invocation),
-        executionQueueDelegate(*this) {}
+
+
+public:
+  BuildSystemFrontendDelegateImpl(llvm::SourceMgr& sourceMgr)
+      : sourceMgr(sourceMgr), executionQueueDelegate(*this) {}
+};
+
+
+
+#pragma mark - BuildSystemFrontendImpl
+
```

Comment:
This is the core of the BuildSystemFrontend refactor.   Merges the state logic into one implementation object and protects that state with a mutex.

Reply:
Perfect